polyphemus/lib/etls/redcap/lib/magma_models.rb

```diff
@@ -0,0 +1,66 @@
+module Redcap
+  class MagmaModelsWrapper
+    attr_reader :project_name, :magma_client
+    def initialize(magma_client, project_name)
+      @project_name = project_name
+      @magma_client = magma_client
+
+      # Fetch existing records separately, since tables
+      #   don't have identifiers and including
+      #   "record_names": "all", "attribute_names": "identifiers"
+      #   will not return their
+      #   templates.
+      @magma_models = magma_client.retrieve(
+        Etna::Clients::Magma::RetrievalRequest.new(
+          project_name: project_name
+        )
+      ).models
+    end
+
+    def models
+      @magma_models
+    end
+
+    def record_ids_for_model(model_name, model_records)
+      # Fetches the relevant record ids for the given model.
+      # For a non-table model, just requests the records directly.
+      # For a table model, requests the parent's record names and then
+      #   returns all the temporary ids from model_records
+      #   that link to the parent records.
+      return model_record_names(model_name) unless is_table?(model_name)
+
+      parent_record_names = model_record_names(parent_model_name(model_name))
+
+      model_records.select do |temp_id, record|
+        parent_record_names.include?(record[parent_model_name(model_name).to_sym])
+      end.keys
+    end
+
+    def parent_model_name(model_name)
+      models.model(model_name.to_s).template.parent
+    end
+
+    def is_table?(model_name)
+      # From a set of magma models, can only tell if a model
+      #    is a table from its parent's definition of the
+      #    child attribute.
+      models.model(
+        parent_model_name(model_name)
+      ).template.attributes.attribute(
+        model_name.to_s
+      ).attribute_type == Etna::Clients::Magma::AttributeType::TABLE
+    end
+
+    def model_record_names(model_name)
```

Comment:
Does make me ponder, though, that eventually we'd consider wanting per-model definition of what constitutes redcap "owned" records.  For instance, imagine assays being split between redcap and other sources.  I think the model class would want to have a hook to say "this is the filter to use when retrieving to find my records".

Reply:
Hm...is that more true at the attribute level than at the individual record level? You're saying, what if `sc_rna_seq` record 1, `date` was from REDCap but record 2 `date` is from a different source? 

I can see "sc_rna_seq" has some subset of attributes from REDCap and others from some other source, but that's controlled by the config JSON, right?

Reply:
Having assay records for a single model come from multiple sources would also imply that there are multiple processing pipelines, which seems unlikely?

Reply:
:shrug:  Ok if it's not likely we don't have to address it.  I fear one day it might occur though because, hey, people do things lol.  But we can also address that on the day that it does occur, it isn't a priority as much as a thought.

Reply:
If you can imagine it, it'll probably happen. But since we don't have a way to pull that metadata info "per record" now, I think we'll have to punt...